projects/cli-files/go-cat/cmd/root.go

```
@@ -0,0 +1,66 @@
+package cmd
+
+import (
+	"bufio"
+	"fmt"
+	"io"
+	"os"
+)
+
+func Execute() {
+	args := os.Args[1:]
+
+	for _, arg := range args {
+		err := checkArgs(arg)
+		if err != nil {
+			fmt.Fprintln(os.Stderr, err)
+			return
+		}
+		fileLines, err := readFileLines(arg)
+		if err != nil {
+			fmt.Fprintln(os.Stderr, err)
+		}
+		for _, fileLine := range fileLines {
+			fmt.Print(fileLine)
+		}
+	}
+}
+
+func checkArgs(arg string) error {
+	if arg == "" {
+		return fmt.Errorf("go-cat: no file specified")
+	}
+	fileInfo, err := os.Stat(arg)
+	if err != nil {
+		return fmt.Errorf("go-cat: '%s': no such file or directory", arg)
```

The error here isn't necessarily that the file doesn't exist - it could be e.g. that you don't have permission to stat the file. It's generally better to report the original error (optionally with some wrapping text), or to return a generic error ("could not open file"), rather than replacing a specific error with another specific error which may not be quite accurate :)

---

If we have to use the original error , can we still test only the case when the file does not exist, right ?

---

Yeah, if we wanted to test other cases we'd either need to be able to trigger them (e.g. make a file we don't have permission to stat), or have a way of injecting a failure (e.g. make a new interface like `FileStater` and pass one as an argument, and in a test we can inject a fake `FileStater` which returns custom errors).